### PR TITLE
RGB LEDs as state indicators with animations, cup illumination, etc 

### PR DESCRIPTION
**Pull request by genkigenki**
Added FastLED library, and LED handling for Neopixels with e.g. WS2811 chip.
LoopLED will be called every 80ms for relatively smooth LED animations.
Number of LEDs and chip type is freely configurable.
Library is using Hardware features of ESP32 to avoid interrup conflicts
2 LED behaviours added to select from: Basic & Plus

LED behaviour "Basic" 
Perfect temp -> Green, one LED in back of machine
Backflush -> white
After coffee is brewed -> white 
Error -> red

LED behaviour "Plus" :
PINon (not standby) -> low brightness white
Perfect temp -> Green, one LED in back of machine
Brew/Steam/Backflush -> white
After coffee is brewed -> White light to show ready coffee for 15s, for the last 10% of the time, smoothly dim down
Initialize, cold start Heat up -> Rainbow (only if configured)
Error message -> Red heartbeat (Water empty, sensor error, etc) 
Steam-heat-up -> yellow

### CODE REVIEW

**Review by LoQue90 — CHANGES_REQUESTED** [review]

File: src/main.cpp [hunk 1]
```diff
@@ -37,6 +37,16 @@
...
 
+//LEDs, LED[0] is the main one indicating also the correct temperature
+#include <FastLED.h>
+CRGB leds[NUM_LEDS];
+unsigned long previousMillisLED;  // initialisation at the end of init()
+const unsigned long intervalLED = 80; // to have slightly smooth animated LED colors
+unsigned int cycleLED = 0; // counter to 255 for color cycling
+unsigned long brewFinishedLEDon = 0; // counter to have LED on to show coffee is ready
+#define BREW_FINISHED_LEDON_DURATION 320 // Time to illuminate the ready coffee cup - 320*intervalLED = 29s
+#define PIDON_LED_BRIGHTNESS 50
```
[the remark this answers is not in the document]
Comment (genkigenki) [reply src/main.cpp:48]: Userconfig - its already quite a few settings for the LEDs, would keep that standard.
Defaults.h would be possible - but to me it seems these are the values which will be saved and are possible to change via web interface?

File: src/StatusLEDvoid.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        // Backflush -> White
+        if (machineState == kBackflush) {
+            fill_solid(leds, NUM_LEDS, CRGB::White); 
+        }        
+
+        // After coffee is brewed -> White light to show ready coffee for ~15s
+        if (brewFinishedLEDon > 0) {
+            brewFinishedLEDon--;
+            // for the last 10% of the time, smoothly dim down white light on cup
+            brightnessLED = (brewFinishedLEDon > (BREW_FINISHED_LEDON_DURATION / 10) ? 200 : 150*brewFinishedLEDon*10/BREW_FINISHED_LEDON_DURATION+PIDON_LED_BRIGHTNESS);
```
[the remark this answers is not in the document]
Comment (genkigenki) [reply src/StatusLEDvoid.h:98]: OK!

File: src/StatusLEDvoid.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            fill_solid(leds, NUM_LEDS, CHSV(0, 0, brightnessLED)); 
+        }
+        
+        // Error message: Red (Water empty, sensor error, etc) - overrides all other states in LED color
+        if (!waterFull || machineState == kSensorError || machineState ==  kEepromError || machineState ==  kEmergencyStop) {
+            if (LED_BEHAVIOUR == 2) {
+                // [For LED_BEHAVIOUR 2]: Red heartbeat
+                fill_solid(leds, NUM_LEDS, CHSV(0, 250, cubicwave8(cycleLED*4))); //Hue of 0 is red
```
[the remark this answers is not in the document]
Comment (genkigenki) [reply src/StatusLEDvoid.h:106]: OK!

File: src/brewvoid.h [hunk 4]
```diff
@@ -344,6 +347,12 @@ void brew() {
                     brewDetected = 0;  // rearm brewDetection
                     brewCounter = kBrewIdle;
                     timeBrewed = 0;
+
+                    //LED illuminates coffee
+                    if (brewFinishedLEDon==0)
```
[the remark this answers is not in the document]
Comment (genkigenki) [reply src/brewvoid.h:359]: OK

File: src/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (LoQue90) [reply src/main.cpp:48]: If you program it to be configured via web interface, standard values should be then in defaults.h

[the remark this answers is not in the document]
Comment (genkigenki) [reply src/main.cpp:48]: Wouldnt configure it there. 
If i just want to keep it constant would it be OK like that?

[the remark this answers is not in the document]
Comment (LoQue90) [reply src/main.cpp:48]: That was my first question, if this is something that the user should be able to configure. A brightness is something that I would like to set myself. 
It's just my thoughts about that.

[the remark this answers is not in the document]
Comment (genkigenki) [reply src/main.cpp]: OK good idea, made the "machine on" brightness configurable in userconfig.

**Pull request closed without merging**